File: src/main/scala/Token.scala
package cql

enum TokenType:
  // Single-character tokens.
  case PLUS, COLON, AT, LEFT_BRACKET, RIGHT_BRACKET,

    // Literals.
    STRING, NUMBER, QUERY_OUTPUT_MODIFIER_KEY, QUERY_FIELD_KEY, QUERY_VALUE,

    // Keywords.
    AND, OR, EOF

case class Token(
    tokenType: TokenType,
    lexeme: String,
    literal: Option[String],
    start: Int,
    end: Int
):
  override def toString = s"${tokenType} ${lexeme} ${literal} ${start}-${end}"

object Token:
  val reservedWords = Map(
    "AND" -> TokenType.AND,
    "OR" -> TokenType.OR
  )

File: prosemirror-client/src/lang/token.ts
export type TokenType = keyof typeof TokenType;

export const TokenType = {
  // Single-character tokens.
  PLUS: "PLUS",
  COLON: "COLON",
  AT: "AT",
  LEFT_BRACKET: "LEFT_BRACKET",
  RIGHT_BRACKET: "RIGHT_BRACKET",

  // Literals.
  STRING: "STRING",
  NUMBER: "NUMBER",
  QUERY_OUTPUT_MODIFIER_KEY: "QUERY_OUTPUT_MODIFIER_KEY",
  QUERY_FIELD_KEY: "QUERY_FIELD_KEY",
  QUERY_VALUE: "QUERY_VALUE",

  // Keywords.
  AND: "AND",
  OR: "OR",
  EOF: "EOF",
} as const;

export class Token {
  public static reservedWords = {
    AND: TokenType.AND,
    OR: TokenType.OR,
  } as const;

  constructor(
    public tokenType: TokenType,
    public lexeme: string,
    public literal: string | undefined,
    public start: number,
    public end: number
  ) {}
  public toString() {
    return `${this.tokenType} ${this.lexeme} ${this.literal} ${this.start}-${this.end}`;
  }
}
